Objective-C/CBLProperties.mm

```diff
@@ -146,97 +145,403 @@ - (BOOL) containsObjectForKey: (NSString*)key {
+                    [oldSubdoc invalidate];
+                }
+            }
+        }
+    }
 }
 
 
-- (void) markChanges {
-    self.hasChanges = YES;
+#pragma mark - PRIVATE: SUBDOCUMENT
+
+
+- (CBLSubdocument*) subdocumentWithRoot: (nullable FLDict)dict
+                         withProperties: (nullable NSDictionary*)props
+                                 forKey: (NSString*)key {
+    CBLSubdocument* subdoc = [[CBLSubdocument alloc] initWithParent: self root: dict];
+    [self attachSubdocument: subdoc forKey: key];
+    if (props)
+        subdoc.properties = props;
+    return subdoc;
+}
+
+
+- (void) attachSubdocument: (CBLSubdocument*)subdoc forKey: (NSString*)key {
+    Assert(!subdoc.parent || subdoc.parent == self,
+           @"Subdocument has already been set to a document or a subdocument.");
+    if (!subdoc.parent)
+        subdoc.parent = self;
+    [subdoc setOnMutate: [self onMutateBlockForKey: key]];
+}
+
+
+- (CBLOnMutateBlock) onMutateBlockForKey: (NSString*)key {
+    __weak typeof(self) weakSelf = self;
+    return ^{
+        __strong typeof(self) strongSelf = weakSelf;
+        [strongSelf markChangedKey: key];
+    };
 }
 
 
 #pragma mark - PRIVATE: FLEECE
 
 
 - (FLValue) fleeceValueForKey: (NSString*) key {
-    return FLDict_GetSharedKey(_root, CBLStringBytes(key), [self sharedKeys]);
+    return FLDict_GetSharedKey(_root, CBLStringBytes(key), self.sharedKeys);
+}
+
+
+- (NSDictionary*) fleeceRootToProperties: (FLDict)root asJson: (BOOL)asJson {
+    int count = FLDict_Count(root);
+    NSMutableDictionary* result = [NSMutableDictionary dictionaryWithCapacity: count];
+    if (count > 0) {
+        FLDictIterator iter;
+        FLDictIterator_Begin(root, &iter);
+        do {
+            NSString* key = [self fleeceValueToKeyString: FLDictIterator_GetKey(&iter)];
+            result[key] = [self fleeceValueToObject: FLDictIterator_GetValue(&iter)
+                                             forKey: key
+                                             asJson: asJson];
+        } while (FLDictIterator_Next(&iter));
+    }
+    return result;
 }
 
 
-- (id) fleeceValueToObject: (FLValue)value {
-    return FLValue_GetNSObject(value, [self sharedKeys], nil);
+- (id) fleeceValueToObject: (FLValue)value forKey: (NSString*)key asJson: (BOOL)asJson {
+    if (!asJson) {
+        FLValueType type = FLValue_GetType(value);
+        switch (type) {
+            case kFLDict: {
+                // TODO: Support Blob
+                return [self subdocumentWithRoot: (FLDict)value withProperties: nil forKey:key];
+            }
+            case kFLArray: {
+                FLArray array = FLValue_AsArray(value);
+                int count = FLArray_Count(array);
+                NSMutableArray* result = [[NSMutableArray alloc] initWithCapacity: count];
+                if (count > 0) {
+                    FLArrayIterator iter;
+                    FLArrayIterator_Begin(array, &iter);
+                    do {
```

Comment:
Actually there is an if check if count > 0 then doing the loop, so I think I will keep the code the same as it looks cleaner than checking if FLArrayIterator_GetValue return nullprt.